### ISSUE TITLE
Regression in FirstRecompileTime windows_chrome_dev_mode benchmark

### ISSUE THREAD

**Issue by kf6gpe** [flutter/flutter#73176]
https://flutter-flutter-perf.skia.org/e/?begin=1608575043&end=1609380055&keys=Xf6d1b74e42baa0f1a8aca1b5fa048e84&requestType=0&xbaroffset=22968

Looks like it's the engine roll at https://github.com/flutter/flutter/pull/72954.

/cc @chinmaygarde this week's sheriff. I don't expect any work on this to happen until after the end of the New Year's holiday, however.

**Comment by zanderso** [comment]
Nothing much in that engine roll that would affect a web benchmark other than the Dart roll: https://github.com/flutter/engine/pull/23281. That Dart roll is also fairly small, but has a change to dart2js that could be relevant: https://github.com/dart-lang/sdk/commit/65110c34486affba9854610d4d20488f2aae432a @rakudrama @grouma

**Comment by kf6gpe** [comment]
@grouma @sigmundch Gentle ping from the Flutter team as part of triage --- any thoughts? Thanks!

**Comment by grouma** [comment]
I don't have much insight into the compiler changes but I have pinged the team to take a look at this. Are there any specific changes that I made that I should look at?

**Comment by rakudrama** [comment]
The [dart2js compile benchmarks](https://golem.corp.goog/Benchmark?team=dart2js&benchmark=CompileDeltaBlueMinify#targets%3Ddart2js-compile%2Cdart2js-null-compile%2Cdart2js-null-production-compile%2Cdart2js-production-compile%2Cdart2js-x-compile%2Cdart2js-x-production-compile%3BmachineTypes%3Dlinux-x64%3Bcpus%3DIntel%20Xeon) show a 0.3% regression for https://github.com/dart-lang/sdk/commit/52db62dd65284192755906691baf66c47a43f47e

/cc @johnniwinther

**Comment by rakudrama** [comment]
How do I find out what the benchmark result is actually testing? dart2js doesn't do much in 50ms - the system `.dill` file takes longer to load, so I'm not sure dart2js is being called.
How do I 'zoom out' in the chart to see longer-term trends and result from the intervening two weeks?

I can test-revert the LinkedHashMap change. How long until we know if it is responsible?
(We do want this change. though, it reduces the heap size of a compile and is part of a project to get larger apps to compile in less memory.)

**Comment by zanderso** [comment]
@jonahwilliams Do you know anything about this benchmark, or where to route the questions from @rakudrama ?

**Comment by jonahwilliams** [comment]
This is a ddc/frontend_server benchmark, so dart2js changes are not likely the cause.

**Comment by jonahwilliams** [comment]
All of the devicelab tests can be run locally, following https://github.com/flutter/flutter/tree/master/dev/devicelab#running-tests-locally .

In this case, you might just need 

```
cd flutter/dev/devicelab
dart bin/run.dart -t windows_chrome_dev_mode
```

This test requires a windows box, but there are linux and macos versions too

**Comment by jonahwilliams** [comment]
@kf6gpe did this regress on the other platforms?

**Comment by kf6gpe** [comment]
No, it did not.
Ray Rischpater, KF6GPE | kf6gpe.org | @kf6gpe
<http://www.twitter.com/kf6gpe>

**Comment by nshahan** [comment]
> did this regress on the other platforms?

> No, it did not.

@kf6gpe @jonahwilliams Just to clarify, what does platforms mean in this context? Does this mean the benchmark is running in dev mode (ddc) in chrome on linux and mac and they show no regressions?

**Comment by zanderso** [comment]
This is a P0, and the range of commits that could have caused it is very narrow. Is it possible to land a revert so that we can close this out, and do a deeper investigation before relanding? Assigning to @sigmundch for re-assignment to the right owner.

**Comment by sigmundch** [comment]
@srujzs 

Based on the roll details:
> 2020-12-24 skia-flutter-autoroll@skia.org Roll Dart SDK from f79fb10acca8 to 1c14d9bb3528 (2 revisions) (flutter/engine#23281)

Seems like we are basically talking about these changes:

https://github.com/dart-lang/sdk/compare/f79fb10acca8...1c14d9bb3528

Only 1 change affects DDC in this range: https://github.com/dart-lang/sdk/commit/bc43d064794e657a15c72554ce0ccb8f63769517

I am however surprised to see this only affecting windows and not any other platform if that were the source of the problem.

I'd be fine trying to revert to quickly verify whether that addresses the issue.

We don't have a windows box setup easily on our end - are there any we can remoteconnect into to investigate this faster? Is it possible to send separate requests to the existing benchmark bots to compare with a patch that reverts this?

**Comment by srujzs** [comment]
I've uploaded a revert here: https://dart-review.googlesource.com/c/sdk/+/178260. It doesn't seem like there's a bot to test that this behavior is changed so the revert will have to land before we can determine if it made a difference.

Is it possible that the preprocessing of native libraries in `pkg/dev_compiler/lib/src/kernel/target.dart` is what's causing the perf regression?

**Comment by zanderso** [comment]
Starting at skiaperf a bit, the benchmark is noisier on macos, but the regression also looks like it might be happening there as well: https://flutter-flutter-perf.skia.org/e/?begin=1607471585&end=1610089956&keys=X6b538ffd72d92c3ae216bdf51f7e9c95&requestType=0

**Comment by sigmundch** [comment]
likewise, that would be my only guess, but honestly, I would be very surprised if that's the case.

I'd like to better understand what's this benchmark measuring. Does it also account for reload/rendering time? what other pieces of the architecture could be at play here?

**Comment by jonahwilliams** [comment]
I ran the numbers locally on my windows machine and came up with:

~46-47 ms first recompile time after the change
~44-45 ms before the change

So despite my computer being fairly powerful,  the timing is consistent before this change. After the engine roll, I don't see as dramatic of a slowdown.

I'm currently building a local engine with a revert of these changes patched in to double check some more numbers

**Comment by jonahwilliams** [comment]
I ran a local engine build with this change applied and reverted:

Local Engine w revert:
52, 52, 54

Local Engine:
56, 52, 53, 52

TLDR: 🤷🏻‍♂️ ?

@zanderso @sigmundch @kf6gpe

We could proceed with the revert & drop the priority. Then follow up once that rolls into the framework and monitor the performance. Unfortunately it sounds like this is a fairly useful feature we're losing, and we don't really have a good way to test the performance first.

Its also possible this is an artifact of the particular bot. When run on my relatively fast computer, I only notice a ~1ms regression.

**Comment by zanderso** [comment]
@ferhatb it sounds like the small-ish regression here is due to new checks that will to make sure the JS interop members are used correctly, providing additional safety, and wanted to hear your input on whether to accept the regression.

**Comment by ferhatb** [comment]
I think it is reasonable for DDC given extra checks. I expect these checks to not affect dart2js release binary with --O4. As long as thats the case, accept.

**Comment by sigmundch** [comment]
Thanks @jonahwilliams for trying it locally. 

FWIW - the revert landed at master, so it should be in the next roll and we should be able to confirm whether the perf regression is really caused by this. I'd like to verify if that's the case, and if so, we can look at a more efficient way to go about it if we think that's worthwhile. It is just surprising that we only notice this on windows.

One thing that we are really surprised to see is that you have a big gap in your local engine at master - maybe the js-interop change had a small regression and a separate change added more and because of the noise in the plot we may be misattributing the regression to the JS-interop change?

**Comment by jonahwilliams** [comment]
Generally the local engine builds are slower since I build them unopt.

**Comment by sigmundch** [comment]
Another data-point, it appears that the performance on this benchmark was previously higher earlier in November:
https://flutter-flutter-perf.skia.org/e/?begin=1606377600&end=1608019200&queries=branch%3Dmaster%26sub_result%3DFirstRecompileTime%26test%3D~windows_chrome_dev_mode&requestType=0&showZero=false

The numbers look similar to the regression after this change. It seem to have gone down on an unrelated change on 12/13 (https://github.com/flutter/flutter/commit/f35238c8041d22344ea14cb90f4ef5e5ce2e9995). So maybe this could be noise in the benchmark?

**Comment by ferhatb** [comment]
Given 1 month trend moving this to P2.

**Comment by srujzs** [comment]
https://flutter-flutter-perf.skia.org/e/?begin=1608575043&end=1610697600&queries=sub_result%3DFirstRecompileTime%26test%3Dwindows_chrome_dev_mode&requestType=0&showZero=false&xbaroffset=22968

Looking at the FirstRecompileTime again, it seems to have gone down back to its earlier time. Looking at the timing of the different runs, something went in on the 10th that improved the performance. Considering that the [roll of the Dart SDK](https://github.com/flutter/engine/commit/88ee6047485a2f89a028c5c3679482e7540b8b86) was committed during that period, and that roll includes the revert of the above CL, the above CL is very likely the suspect.

There still is no change in linux and macOS times, so it's not evident why this is Windows only. It also isn't evident why it happens on the bot but isn't measured locally.

**Comment by Hixie** [comment]
Does that mean this is resolved? At least in terms of the regression (and thus its P2 nature), if not maybe understanding what caused it?

**Comment by srujzs** [comment]
Yes, the regression is resolved. My only worry is potentially reintroducing the regression as I attempt to fix the original CL, which is why I'm not 100% on closing this issue, although we can certainly reopen as needed. If there's any concrete way to test the bot without submitting the fix and monitoring to revert again, please let me know.

**Comment by Hixie** [comment]
You can run the test locally, in principle, and compare the data before and after.

You are welcome to keep the bug open if that helps tracking this though (e.g. to track relanding what was reverted). I'll remove the P2 label in the meantime. Please don't hesitate to update the bug title as necessary to describe what you desire from this bug. Thanks!

**Comment by srujzs** [comment]
So, there's a bit of noise that was introduced a day or two before my revert of the revert (with a fix) [landed](https://github.com/dart-lang/sdk/commit/787ec114e1288ee63122c5ee7bb62b939d833b08):https://flutter-flutter-perf.skia.org/e/?begin=1608575043&end=1612080000&queries=sub_result%3DFirstRecompileTime%26test%3Dwindows_chrome_dev_mode&requestType=0&showZero=false&xbaroffset=22968

However, accounting for that noise, there is no diff between before and after this landed in the FirstRecompileTime for windows_chrome_dev_benchmark, so this is resolved. I unfortunately don't think I have permissions to close this issue (I don't have a close with comment option), so @Hixie, can you please close this? Thanks!
